cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/cluster/KafkaClusterOperations.java

```diff
@@ -230,16 +289,49 @@ public void create(String namespace, String name, Handler<AsyncResult<Void>> han
         }
     };
 
+    private final CompositeOperation<TopicController> deleteTopicController = new CompositeOperation<TopicController>() {
+
+        @Override
+        public Future<?> composite(String namespace, ClusterOperation<TopicController> clusterOp) {
+            TopicController topicController = clusterOp.cluster();
+            Future<Void> fut = deploymentOperations.delete(namespace, topicController.getName());
+            return fut;
+        }
+
+        @Override
+        public ClusterOperation<TopicController> getCluster(String namespace, String name) {
+            Deployment dep = deploymentOperations.get(namespace, TopicController.topicControllerName(name));
+            return new ClusterOperation<>(TopicController.fromDeployment(namespace, name, dep), null);
+        }
+    };
+
     @Override
     protected void delete(String namespace, String name, Handler<AsyncResult<Void>> handler) {
-        execute(CLUSTER_TYPE_KAFKA, OP_DELETE, namespace, name, deleteKafka, ar -> {
-            if (ar.failed()) {
-                handler.handle(ar);
+
+        // first check if the topic controller was really deployed
+        ClusterOperation<TopicController> clusterOp = deleteTopicController.getCluster(namespace, name);
+        if (clusterOp.cluster() != null) {
+            execute(CLUSTER_TYPE_TOPIC_CONTROLLER, OP_DELETE, namespace, name, deleteTopicController, topicControllerDone -> {
+                if (topicControllerDone.failed()) {
+                    handler.handle(topicControllerDone);
+                } else {
+                    deleteKafkaAndZookeeper(namespace, name, handler);
```

Comment:
Why is the delete function the only one which looks completely different? Create and update operate over all three main resources ... but delete works only on topic controller and the rest is done by other method.

Reply:
@scholzj because we delete the topic controller first, but there might not be a topic controller, so the logic for delection of kafka and zk clusters would be duplicated between the two branches. 

An alternative would be to use a dummy topic controller cluster if a tc was not configured, whose creation, deletion and update were noops. I think @ppatierno rejected that because it would make the logging confusing, saying it was creating a tc that wasn't configured, but maybe we should reconsider. Or use futures and `compose()` them for sequential execution.

Reply:
I really don't understand why all these ideas for making the code more complex.